index/scorch/snapshot_rollback.go

```diff
@@ -0,0 +1,183 @@
+//  Copyright (c) 2017 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scorch
+
+import (
+	"bytes"
+	"log"
+
+	"github.com/blevesearch/bleve/index/scorch/segment"
+)
+
+type indexSnapErr struct {
+	i *IndexSnapshot
+	e error
+}
+
+// Snapshot fetches the latest persisted IndexSnapshot.
+func (s *Scorch) Snapshot() (*IndexSnapshot, error) {
+	if s.rootBolt == nil {
+		return nil, nil
+	}
+
+	ch := make(chan *indexSnapErr)
+
+	go func() {
+		// start a read-only transaction
+		tx, err := s.rootBolt.Begin(false)
+		if err != nil {
+			ch <- &indexSnapErr{nil, err}
+			return
+		}
+
+		snapshots := tx.Bucket(boltSnapshotsBucket)
+		if snapshots == nil {
+			ch <- &indexSnapErr{nil, nil}
+			return
+		}
+
+		c := snapshots.Cursor()
+		for k, _ := c.Last(); k != nil; k, _ = c.Prev() {
+			_, snapshotEpoch, err := segment.DecodeUvarintAscending(k)
+			if err != nil {
+				log.Printf("Snapshot: unable to parse segment epoch %x, continuing", k)
+				continue
+			}
+
+			snapshot := snapshots.Bucket(k)
+			if snapshot == nil {
+				log.Printf("Snapshot: snapshot key, bucket bucket missing %x, continuing", k)
+				continue
+			}
+
+			indexSnapshot, err := s.loadSnapshot(snapshot)
+			if err != nil {
+				log.Printf("Snapshot: unable to load snapshot, %v, continuing", err)
+				continue
+			}
+			indexSnapshot.epoch = snapshotEpoch
+
+			ch <- &indexSnapErr{indexSnapshot, nil}
+			return
+		}
+
+		ch <- &indexSnapErr{nil, nil}
+	}()
+
+	rv := <-ch
+
+	return rv.i, rv.e
+}
+
+// SnapshotPrevious returns the next older, previous IndexSnapshot
+// based on the provided IndexSnapshot, allowing the application to
+// walk backwards into the history of a store at previous points in
+// time. A nil returned snapshot means no previous snapshots are
+// available.
+func (s *Scorch) SnapshotPrevious(is *IndexSnapshot) (*IndexSnapshot, error) {
```

Comment:
Same here with this method name?  Perhaps PreviousPersistedSnapshot()?

Also, this code here, at a first glance, looks strikingly similar to the new Snapshot() method that you've proposed above?  Maybe there's a way to consolidate, and only offer a single new method?
  Let's try... PreviousPersistedSnapshot(is *IndexSnapshot), but it allows the callers to pass in a nil value, which means "just give me the last persisted snapshot, if any"?

Reply:
Done.